libs/knowledge-store/ragstack_knowledge_store/base.py

```diff
@@ -0,0 +1,345 @@
+from __future__ import annotations
+
+import itertools
+from abc import abstractmethod
+from typing import (
+    Any,
+    AsyncIterable,
+    ClassVar,
+    Collection,
+    Iterable,
+    Iterator,
+    List,
+    Optional,
+    Union,
+)
+
+from langchain_core.callbacks import (
+    AsyncCallbackManagerForRetrieverRun,
+    CallbackManagerForRetrieverRun,
+)
+from langchain_core.documents import Document
+from langchain_core.load import Serializable
+from langchain_core.runnables import run_in_executor
+from langchain_core.vectorstores import VectorStore, VectorStoreRetriever
+from pydantic import Field
+
+
+class Node(Serializable):
+    """Node in the KnowledgeStore graph"""
+
+    id: Optional[str]
+    """Unique ID for the node. Shall be generated by the KnowledgeStore if not set"""
+    metadata: dict = Field(default_factory=dict)
+    """Metadata for the node. May contain information used to link this node 
+    with other nodes."""
+
+
+class TextNode(Node):
+    text: str
+    """Text contained by the node"""
+
+
+def _texts_to_nodes(
+    texts: Iterator[str],
+    metadatas: Iterator[dict],
+    ids: Iterator[str],
+) -> Iterator[Node]:
+    for text, metadata, node_id in zip(texts, metadatas, ids):
+        yield TextNode(
+            id=node_id,
+            metadata=metadata,
+            text=text,
+        )
+
+
+def _documents_to_nodes(
+    documents: Iterator[Document], ids: Iterator[str]
+) -> Iterator[Node]:
+    for (
+        doc,
+        node_id,
+    ) in zip(documents, ids):
+        yield TextNode(
+            id=node_id,
+            metadata=doc.metadata,
+            text=doc.page_content,
+        )
+
+
+class KnowledgeStore(VectorStore):
```

Comment:
I think we'll want to iterate on naming (`KnowledgeStore`, `ContentStore`, `DocumentStore`, etc.) have all come up. But that should be in a separate PR (don't want to block the functionality improvements in other PRs which are going to be used in upcoming posts, etc.)